### ISSUE TITLE
[ASSETS] MigrateFileTask runs out of memory 

### ISSUE THREAD

**Issue by torleif** [silverstripe/silverstripe-framework#8664]
## Affected Version

Framework / assets 4.3.0-RC1

## Description

I have a site with gigabytes of files, and running ```vendor/silverstripe/framework/sake dev/tasks/MigrateFileTask``` causes out of memory errors. Running it returns errors such as ```Fatal Error (E_ERROR): Allowed memory size of 4487905280 bytes exhausted (tried to allocate 28800001 bytes)```

I've upped the memory limit to 4 gigs to try and get through the task.

It seems it is progressing each time it's run, as the byte allocation is different.

Ideally there wouldn't be a memory leak so the entire task can run without running out of memory.

## Steps to Reproduce

Run ```vendor/silverstripe/framework/sake dev/tasks/MigrateFileTask ``` with a large asset base

**Comment by chillu** [comment]
We actually have an issue to determine rough maximums for this task: https://github.com/silverstripeltd/open-sourcerers/issues/45

Out of interest, do you have very large images (>4000px dimensions) in your site? How many files do you have in total, and how many are processed before it runs out of memory? Just trying to narrow down the use case.  4GB should definitely be enough to run this, I think 512MB is a reasonable max that should allow most environments to run this task.

**Comment by torleif** [comment]
There's 23.0GB of files, and 197,667 files in total. I'm imagining there's a few large images over 4000px in there. 

It seems to fault on the second part of the task, generating thumbnails. it gets to about 68k in the list of files.

**Comment by wernerkrauss** [comment]
@torleif could this be related to https://github.com/silverstripe/silverstripe-framework/issues/8570 ? Dunno if that patch is available in 4.3RC-1 or was merged later.

What if you increase your memory limit manually in `FileMigrationHelper::run()`? 

```
ini_set('memory_limit', -1); 
```

helped for me in SS4.2.

**Comment by torleif** [comment]
I'm overwrote the memory limit in the Envornment file, to see if it would work with that. It looks like it was setting the limit to the default PHP amount. 

after running that I'm getting a new error: 

```
VirtualAlloc() failed: [0x000005af] The paging file is too small for this operation to complete.


VirtualAlloc() failed: [0x000005af] The paging file is too small for this operation to complete.


Fatal error: Out of memory (allocated 16114515968) (tried to allocate 24576 bytes) in C:\Users\torleifw\op\vendor\intervention\image\src\Intervention\Image\Gd\Decoder.php on line 136

VirtualAlloc() failed: [0x000005af] The paging file is too small for this operation to complete.


Fatal error: Out of memory (allocated 16114515968) (tried to allocate 32768 bytes) in C:\Users\torleifw\op\vendor\monolog\monolog\src\Monolog\Formatter\NormalizerFormatter.php on line 1

```

which looks like "you're running out of memory"

**Comment by lerni** [comment]
Looks like GD? Had the same issue. I've worked around using "imagick" as ImageManager driver. If that works for you as well, it points to something smelly with GD.

**Comment by torleif** [comment]
@lerni I tried swapping to GD, it seems to have very similar memory issues

**Comment by lerni** [comment]
@torleif Doesn't the error above look already like GD? It worked for me with "imagick" but not with "GD"!

**Comment by torleif** [comment]
apologies, I meant to write I swapped to imagick from GD, which resulted in similar errors.

**Comment by maxime-rainville** [comment]
Can any one that got errors tell me what version of PHP they were using? I'm having problem replicating this locally with these parameters:
* PHP7.1
* 700 images with a total of 2100MB.
* My image sizes vary between 1MB and 6MB.
* I capped my memory usage at 100MB
* I'm using the CLI

**Comment by maxime-rainville** [comment]
@ingo Just looking back through the comments and @torleif set up is way beyond our acceptance criteria of 10K files with 2MB JPEG.

> There's 23.0GB of files, and 197,667 files in total. I'm imagining there's a few large images over 4000px in there.

Do we want to revise our acceptance criteria up?

**Comment by chillu** [comment]
Those ACs are taken from our team's DoD. This is mainly about establishing limits on different configurations, and then extrapolating from that. I don't want you to spend a long time generating that amount of data in order to establish a higher maximum. As discussed, please start by talking to SilverStripe Operations about their past experiences.

**Comment by maxime-rainville** [comment]
Had another go at it. The actual memory usage was going over the limit I had set on the CLI. So something must be calling `ini_set('memory_limit')` to increase the minimum amount of memory available.

**Comment by maxime-rainville** [comment]
I found that when you try to get the actual memory usage with `memory_get_peak_usage` you always get something very small. While if you look at the OS resource monitor you get a much higher number. 
In the example below, PHP report's a peak memory usage of 18MB while the UBuntu resource monitor reports 1.4GB for my PHP process.

https://youtu.be/fsJg1RReYmU

https://github.com/silverstripe/silverstripe-asset-admin/pull/904 addresses the memory issues by reverting to using ImageBackendFactory to create the ImageBackend object. This appears to improve garbage collection. OS reported memory usage never goes above 250MB on my test case.

https://youtu.be/vCMvbySoXj0

**Comment by kinglozzer** [comment]
@maxime-rainville Yeah I’d definitely recommend a profiler rather than `memory_get_usage()` / `memory_get_peak_usage()`. I’ve found https://blackfire.io to be good if you don’t want the pain of setting up something like [tideways](https://tideways.com/profiler/xhprof-for-php7)

**Comment by kinglozzer** [comment]
I’m not sure how the Ubuntu resource monitor works, but I’d guess that’s total memory used (allocated, de-allocated, re-allocated, de-allocated etc) rather than the max used at any single point in time, so I don’t think that would cause out-of-memory errors.

**Comment by maxime-rainville** [comment]
@bergice Here's some pointers on reviewing things. I split this one up into many PRs. They don't have any cross dependencies and can all be tested individually.

* https://github.com/silverstripe/silverstripe-framework/pull/8745 is meant to cover "Clearly communicates the status of the task on subsequent runs". I'm not super happy with it, but without considerably altering the script there's not really any elegant way of doing it.
* #8733 is not technically related. It just adds a debug parameter to printout the exec time and memory usage.
* https://github.com/silverstripe/silverstripe-asset-admin/pull/905 improves memory allocation. To be honest, I couldn't figure out why it was coded the way it was. To me it looks like the original developer want of is way to override the default injector behaviour, but didn't really make it clear why. This PR also adds a 9MB limit when loading image file. If your image is bigger than 9MB, the thumbnail generation will be skipped. From my local testing, 9MB is about the point where the memory usage goes over 512MB. Maybe try confirming this on your local environment and on SSP.
* https://github.com/silverstripe/silverstripe-assets/pull/208 simplifies a bit of code. It's doesn't make any difference to execution time or memory allocation.
* https://github.com/silverstripe/silverstripe-framework/pull/8746 Documents how to debug/tweak the file migration task when dealing lots of files. I didn't put hard numbers. That probably will have to wait for your CWP/SSP testing.
